sungrow/integrationpluginsungrow.cpp

```diff
@@ -0,0 +1,375 @@
+            if (m_monitors.contains(thing)) {
+                qCDebug(dcSungrow()) << "Unregister monitor because the setup has been aborted.";
+                hardwareManager()->networkDeviceDiscovery()->unregisterMonitor(m_monitors.take(thing));
+            }
+        });
+
+        QHostAddress address = m_monitors.value(thing)->networkDeviceInfo().address();
+
+        qCInfo(dcSungrow()) << "Setting up Sungrow on" << address.toString();
+        auto sungrowConnection = new SungrowModbusTcpConnection(address, m_modbusTcpPort , m_modbusSlaveAddress, this);
+        connect(info, &ThingSetupInfo::aborted, sungrowConnection, &SungrowModbusTcpConnection::deleteLater);
+
+        // Reconnect on monitor reachable changed
+        connect(monitor, &NetworkDeviceMonitor::reachableChanged, thing, [=](bool reachable){
+            qCDebug(dcSungrow()) << "Network device monitor reachable changed for" << thing->name() << reachable;
+            if (!thing->setupComplete())
+                return;
+
+            if (reachable && !thing->stateValue("connected").toBool()) {
+                sungrowConnection->modbusTcpMaster()->setHostAddress(monitor->networkDeviceInfo().address());
+                sungrowConnection->reconnectDevice();
+            } else if (!reachable) {
+                // Note: Auto reconnect is disabled explicitly and
+                // the device will be connected once the monitor says it is reachable again
+                sungrowConnection->disconnectDevice();
+            }
+        });
+
+        connect(sungrowConnection, &SungrowModbusTcpConnection::reachableChanged, thing, [this, thing, sungrowConnection](bool reachable){
+            qCDebug(dcSungrow()) << "Reachable changed to" << reachable << "for" << thing;
+            if (reachable) {
+                // Connected true will be set after successfull init
+                sungrowConnection->initialize();
+            } else {
+                thing->setStateValue("connected", false);
+                thing->setStateValue(sungrowInverterTcpCurrentPowerStateTypeId, 0);
+
+                foreach (Thing *childThing, myThings().filterByParentId(thing->id())) {
+                    childThing->setStateValue("connected", false);
+                }
+
+                Thing *child = getMeterThing(thing);
+                if (child) {
+                    child->setStateValue(sungrowMeterCurrentPowerStateTypeId, 0);
+                    child->setStateValue(sungrowMeterCurrentPhaseAStateTypeId, 0);
+                    child->setStateValue(sungrowMeterCurrentPhaseBStateTypeId, 0);
+                    child->setStateValue(sungrowMeterCurrentPhaseCStateTypeId, 0);
+                }
+
+                child = getBatteryThing(thing);
+                if (child) {
+                    child->setStateValue(sungrowBatteryCurrentPowerStateTypeId, 0);
+                    child->setStateValue(sungrowBatteryBatteryLevelStateTypeId, 0);
```

Comment:
Please reset only the live power data which gets evaluated in the energy database. This is important for not having a constant wrong load when a device is disconnected. That does not apply to energy states (always counting up) and stuff like battery SoC.